### ISSUE TITLE
Build bridge.js on the fly (exclude things that are not actually used)

### ISSUE THREAD

**Issue by DaniilVeriga** [bridgedotnet/Bridge#868]
Originally asked here:
http://forums.bridge.net/forum/general/feature-requests/305

A related question:
http://forums.bridge.net/forum/community/help/1468

**Comment by heavenmaster** [comment]
As a support, I would like to express my opinion. This is very important, especially for the users who build their libraries upon bridge.net. Currently Bridge.NET is pretty large comparing to other js libs and growing. Technically it is possible to get known what parts / classes are used by user code and build js with only required parts. However I think it might be not easy task. I heard about 2.0 will use Roslyn, if so 2.0 must have such feature, atleast optional. Thank you!

**Comment by DaniilVeriga** [comment]
@heavenmaster, thank you for sharing your opinion. Yes, it is not an easy task, but this feature is definitely going to appear.

**Comment by Jluil** [comment]
If you could write bridge core in bridge.net this could become a doable task.

**Comment by DaniilVeriga** [comment]
Yeah, eating our own dogs food, i.e. writing Bridge Core using Bridge itself is something that we are keeping in mind for awhile. Hope to see it implemented at some point in the future.

**Comment by geoffreymcgill** [comment]
Several classes are already being compiled from C# into .js, including several native C# classes. 

https://github.com/bridgedotnet/Bridge/tree/master/Transpiled

The Bridge.Collections package is also compiled directly from the native C# source. 

https://github.com/bridgedotnet/Frameworks/tree/master/Collections

More will be coming, eventually all.

**Comment by hardhub** [comment]
Is it ready for now? My minified version of bridge.min takes 390 kb in size.
It is abnormal absolutely... most JS frameworks require only 20-60kb.
So bridge should take 100-150kb maximum.

**Comment by hardhub** [comment]
Non-minified bridge.js is 378 KB (includes all comments and documentation)
    Minified bridge.min.js is 168 KB (comments and documentation removed)
    GZipped bridge.min.js is 49 KB

This is not true for nowadays... it takes much more... 900kb unmimified...

I think it is really good idea to build bridge.json using C#. So step by step we can create bridge file smaller... and finally do all on the fly.

**Comment by geoffreymcgill** [comment]
> Is it ready for now?

The issue is still open.

> most JS frameworks require only 20-60kb.

I have seen js libraries that are bigger than 60kb.

**Comment by hardhub** [comment]
> The issue is still open.
OK, just a hope :)

> I have seen js libraries that are bigger than 60kb.
But as I know mostly each frontend developer has similar mantra about "less size".
Even new frameworks come with the same functionality and less size.

For example minified version for:
Bobril with all modules - 36Kb
KnockoutJS - 22 Kb
Vue2 - 71Kb
Aurelia - 18 kb

I agree it can be so that some frameworks are really big in size. But not all... I see trend to reduce size...

For example size of my last site in production with minified CSS, JS, with images in SVG and with all HTML templates (in other words whole app with all data) requires 560 Kb only.

**Comment by geoffreymcgill** [comment]
> But not all... I see trend to reduce size...

I see the very opposite. 20 years ago I didn't use any JavaScript on my first website, now I use a lot.

Are you directly comparing Bridge to Bobril, KnockoutJS, Vue2 and Aurelia?

**Comment by hardhub** [comment]
> I see the very opposite. 20 years ago I didn't use any JavaScript on my first website, now I use a lot.
And it was tons of HTML without JS. Now everywhere are small JS libs, bundlers, minifiers, HTML templates and components.

> Are you directly comparing Bridge to Bobril, KnockoutJS, Vue2 and Aurelia?
Not sure what you mean directly... From point of view of web developer maybe I can say yes.

What is Bridge.Net for web developer? 
Transpiler to ES5 allowing us to code in C# and manipulate DOM.

What is some popular framework (let's say Aurelia) for web developer?
Transpiler from ES6 to ES5 with support of DOM manipulation (+MVVM, components and so on).
+ reactive programming in knockout
+ virtual DOM in React based family...
and other features..

Bridge good start to remove JS hell... Finally it can end in Bridge to WASM.
And it will be game over for JS development in browsers.

IMHO, web community is already ready to use your idea (or will be ready very soon). So you need offer just good implementation. Is it size important as part of that good implementation? Yes I think so.

**Comment by vladsch** [comment]
Hi @hardhub,

Thanks for your comments.

> What is Bridge.Net for web developer?
Transpiler to ES5 allowing us to code in C# and manipulate DOM.

I cannot agree with you. The main goal of Bridge.NET is repeat .NET core library in another platform (at this moment, it is any platform which supports javascript). It means that, for example, 64-bit types, decimal, LINQ, regular expressions engine, globalization and so on should be available for developers and you can expect that result of transpiled code will be same as for .NET. If you look into source code of .NET core then you will understand that such functionality cannot be implemented in 100kb library. Of course, Bridge doesn't cover many things from .NET yet but it is main goal in future.

I agree that need some approach to customize `bridge.js` file (automatically and manually) to exclude some unused code. We are reviewing some possible ways and I hope that it will be implemented in near releases

**Comment by hardhub** [comment]
OK, nice... Bridge.Net is more than transpiler (as any web framework with routing, MVVM, FRP etc depending on framework)... 
But for example your implementation of decimals is another framework... what can do web developer in case of pure JS implementation? He can connect it separately...
In JS world for globalization momentJS can be used. MomentJS is available with and without locales.

What can be made in Bridge.Net? We can use namespaces...
For example  System.Text.RegularExpressions, if no classes from this namespace are used - do not include it..

In such way you will achieve if not 100kb then let's say 200kb but it will be still amazing result!

**Comment by Zaid-Ajaj** [comment]
Anyone heard of [Rollup](https://github.com/rollup/rollup)? This has great dead-code elimination capabilities that will remove the code that it not used. I don't know how to intergrate it with the Bridge core library. But we are using it already with [fable](https://github.com/fable-compiler/Fable) to import only the used modules which I think is exactly what you guys are trying to do here

**Comment by geoffreymcgill** [comment]
Okay, thanks for the feedback. This has gone long enough; please stop. 

We know all this and are working on some solutions.

**Comment by michaelcheers** [comment]
@geoffreymcgill, @vladsch, with reflection in some cases this may be impossible. I think a better solution might be loading JS as necessary with http requests.

**Comment by hardhub** [comment]
@michaelcheers 
We are trying to avoid it... and use bundling even in case of HTTP2.

So I think something like Rollup is much more preferable.
It will reduce size so much as possible...

Regarding reflection.. I think we can specify bundling settings in some way.
For example like I did that to exclude console module from bridge JS.

**Comment by MarcoTheFirst** [comment]
At my company we develop quite large and complex enterprise single page apps using bridge, and neither we nor our users care if it takes 1 second more to load. In fact, when comparing JS file transfer to the sum of all other transfers at runtime (REST-traffic, HTML loading, CSS loading, images, fonts etc.), then really it doesn't matter. Also when looking at other large web apps, it is unbelievably overwhelming how much JS code they deploy to the browser, and still it works completely fine. LiquidPlanner, our former project planning tool, deploys roughly 3.5 MB of minified JS!

> We follow two rules in the matter of optimization:
> Rule 1: Don't do it.
> Rule 2 (for experts only). Don't do it yet - that is, not until you have a perfectly clear and unoptimized solution.
> _Michael A. Jackson_

**Comment by hardhub** [comment]
@MarcoTheFirst Our app takes 1.4MB with all images, graphics, fonts, code, initial data during first load. It is minified and gzipped size (where applicable to avoid BREACH).
You are at enterprise, but please think about mobile devices on poor 3G.
So for now on good connection whole app can be loaded in up to 2 seconds (no cache). 
Google Search page with minimal content for example takes up to 1 second on the same connection.

Minified Bridge files + app compiled by Bridge take 3MB, zipped ~ 400Kb.
So for now only compression helps us. But reducing even of that zipped size in 2 times still will be appreciated if you take care about speed.

If functionality or size? Functionality!
But size is still matter, and must be considered if functionality is not impacted.
Just my three cents.
